src/Conversion/ONNXToTorch/ONNXToTorchCommon.cpp

```diff
@@ -16,34 +16,80 @@
-    return Torch::StringType::get(stringType.getContext());
-  });
-
-   // Use UnrealizedConversionCast as the bridge so that we don't need to pull in
-   // patterns for other dialects.
-  addSourceMaterialization([&](OpBuilder &builder, Type resultType,
+  /// Unrealized conversion cast
+  auto addUnrealizedCast = [&](OpBuilder &builder, Type resultType,
                                ValueRange inputs,
                                Location loc) -> Optional<Value> {
     if (inputs.size() != 1)
       return llvm::None;
     return builder.create<UnrealizedConversionCastOp>(loc, resultType, inputs)
         .getResult(0);
+  };
+  addSourceMaterialization(addUnrealizedCast);
+  addTargetMaterialization(addUnrealizedCast);
+
+  /// String conversion
+  addConversion([](StringType stringType) {
+    return Torch::StringType::get(stringType.getContext());
   });
 
-  addTargetMaterialization([&](OpBuilder &builder, Type resultType,
-                               ValueRange inputs,
-                               Location loc) -> Optional<Value> {
-    if (inputs.size() != 1)
-      return llvm::None;
-    return builder.create<UnrealizedConversionCastOp>(loc, resultType, inputs)
-        .getResult(0);
+  /// Tensor conversion
+  addConversion([](TensorType type) -> Optional<Type> {
+    return Torch::ValueTensorType::get(
+      type.getContext(), type.getShape(), type.getElementType());
+  });
+  auto addTensorCast = [](OpBuilder &builder, TensorType type,
+                          ValueRange inputs, Location loc) -> Value {
+    assert(inputs.size() == 1);
+    assert(inputs[0].getType().isa<Torch::BaseTensorType>());
+    return builder.create<torch::TorchConversion::ToBuiltinTensorOp>(loc, inputs[0]);
+  };
+  addSourceMaterialization(addTensorCast);
+  addArgumentMaterialization(addTensorCast);
+  addTargetMaterialization([](OpBuilder &builder, Torch::ValueTensorType type,
+                              ValueRange inputs, Location loc) -> Value {
+    assert(inputs.size() == 1);
+    assert(inputs[0].getType().isa<TensorType>());
+    return builder.create<torch::TorchConversion::FromBuiltinTensorOp>(
+      loc, type, inputs[0]);
+  });
+
+  /// Create tensor to value tensor type conversion and ensure that
+  /// we always use signed integer types. i64 -> si64 conversion.
+  addConversion([](RankedTensorType type) -> Type {
+    Type elementType = type.cast<TensorType>().getElementType();
+    if (type.getElementType().isSignlessInteger(64)) {
+      elementType = IntegerType::get(type.getContext(), 64, IntegerType::Signed);
+    }
+    return Torch::ValueTensorType::get(type.getContext(),
+      type.getShape(), elementType);
+  });
+  addTargetMaterialization([](OpBuilder &builder,
+                              RankedTensorType type, ValueRange inputs,
+                              Location loc) -> Optional<Value> {
+    /// Other builtin integer types could be handled by other materializers.
```

Comment:
In the way you handle this with an if statement below,  this would not work, because you would need to make the materialization fail (return None) to allow another to try again.

Reply:
Added `return llvm::None`; hope I did not miss something again.